include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,423 @@
+{
+  // Check if lower bound is a vector of a single dimension.
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
+  if (upperBound.n_rows == 1)
+    upperBound = upperBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+  std::vector<size_t> shuffle(populationSize);
+
+  // Weight vectors, where each one of them represents a decomposition.
+  std::vector<arma::vec> weights(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    weights[i] = arma::vec(numObjectives, arma::fill::randu);
+    shuffle[i] = i;
+  }
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> weightNeighbourIndices(populationSize, neighbourhoodSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    // To temporarily store the distance between weights(i) and each other weights.
+    arma::vec distances(populationSize);
+    for (size_t j = 0; j < populationSize; j++)
+    {
+      distances(j) = 0;
+      for (size_t w = 0; w < numObjectives ; w++)
+      {
+        distances(j) += std::pow(weights[i](w) - weights[j](w), 2);
+      }
+      distances(j) = std::sqrt(distances(j));
+    }
+    arma::uvec sortedIndices = arma::stable_sort_index(distances, "descend");
+    for (size_t iter = 1; iter <= neighbourhoodSize; iter++)
+      weightNeighbourIndices(i, iter - 1) = sortedIndices(iter);
+  }
+
+  // 1.3 Random generation of the initial population.
+  std::vector<MatType> population(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population[i] = arma::randu<MatType>(iterate.n_rows, iterate.n_cols)
+        - 0.5 + iterate;
+  }
+
+  // 1.3 F-value initialisation for the population.
+  std::vector<arma::vec> FValue(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    FValue[i].resize(numObjectives);
+  EvaluateObjectives(population, objectives, FValue);
+
+  // 1.4 Initialize the ideal point z.
+  std::vector<arma::vec> idealPoint(1);
+  idealPoint[0].resize(numObjectives);
+  std::vector<MatType> iterateWrapper(1);
```

Comment:
Maybe I missed something, but it looks like we don't change the size of `iterateWrapper` and could directly use `MatType`. Same for `idealPoint`.

Reply:
I had to wrap iterate and idealPoint in single element vectors because I was passing them to `EvaluateObjectives `which accepted a vector as arguments. I could alternatively overload the function to take single elements as arguments.

Reply:
We could just update `EvaluateObjectives` to take `arma::mat` instead of `std::vector<arma::vec>` as input, it's an internal optimizer function, so there is no problem and `EvaluateObjectives` currently unpacks the `std::vector` and calls `Evaluate` on that function, so we can just use `Evaluate(.col())` instead.

Reply:
I thought of comparing the performance value of the fronts obtained from MOEA/D and NSGA2, the values for the former were in the correct range but those for the latter were all 0! The least possible values can be 0.98 for Fonseca Fleming function and 2 for Schaffer N1 function.

Reply:
Not sure why, but replacing the for each loop in the end with a normal for loop resulted in expected values of performance (in NSGA2).

Reply:
Can I implement other decomposition approaches later? This one has been going on for quite some time now.:sweat_smile:

Reply:
I don't know why this build is failing, the one on my branch [passed](https://ci.appveyor.com/project/UtR491/ensmallen).

Reply:
Hi @zoq, can you take some time out to review this, please? I thought the appveyor build failing was just a one-off, but it failed twice in a row, I tried to figure out why but could not. Also, [it passed](https://ci.appveyor.com/project/UtR491/ensmallen) on my branch so I am not sure what the problem is.:confused:

Reply:
Hm looks like the force-push removed the build log?

Reply:
Yes. Actually I just dropped the commits that caused the failure. Maybe, it was some timeout issue because both the times, it failed after running for 90 minutes.